### PR TITLE
Proposal for a keyboard, mouse, and human input API

### PR DESCRIPTION
**Pull request by Perksey**
# Summary of the PR
This proposal covers the following areas:
- `Silk.NET.Input`
- `Silk.NET.Input.Common`
- `Silk.NET.Input.Desktop`
- `Silk.NET.Input.Mobile`

# Related issues, Discord discussions, or proposals
None.

# Further Comments
None.

### CODE REVIEW

File: documentation/proposals/Proposal - Input.md [hunk 1]
```diff
@@ -0,0 +1,390 @@
...
+    event Action<IJoystick, Button> ButtonDown;
+    event Action<IJoystick, Button> ButtonUp;
+    event Action<IJoystick, Axis> AxisMoved;
+    event Action<IJoystick, Hat> HatMoved;
+}
+```
+
+### IGamepad
+```cs
+public interface IGamepad : IInputDevice
```
Comment (AzyIsCool) [inline documentation/proposals/Proposal - Input.md]: Maybe also have a ``event Action<IGamepad, Trigger> TriggerMoved`` too?

File: documentation/proposals/Proposal - Input.md [hunk 2]
```diff
@@ -0,0 +1,390 @@
...
+    bool IsButtonPressed(MouseButton btn);
+    event Action<IMouse, MouseButton> MouseDown;
+    event Action<IMouse, MouseButton> MouseUp;
+    event Action<IMouse, ScrollWheel> Scroll;
+}
+```
+
+### IInputDevice
+```cs
+public interface IInputDevice
```
Comment (AzyIsCool) [inline documentation/proposals/Proposal - Input.md]: Maybe if we also have a ``event Action<IInputDevice> Connected`` or change ``event Action<IInputDevice> Disconnection`` to ``event Action<IInputDevice, bool> ConnectionChanged`` here too?

File: documentation/proposals/Proposal - Input.md (hunk 1 above)
Comment (Perksey) [reply documentation/proposals/Proposal - Input.md]: Oh crap, completely forgot about that! Thanks for pointing that out!

File: documentation/proposals/Proposal - Input.md (hunk 2 above)
Comment (Perksey) [reply documentation/proposals/Proposal - Input.md]: Did the latter, thanks!

**Comment by AzyIsCool** [comment]
Umm would a D-Pad on a Gamepad be classed as 4 buttons because Gamepad typically have a D-Pad?

**Comment by Perksey** [comment]
In older versions of this proposal, I did have a `DPad` struct, but instead, I decided to ditch it in favour of the buttons as the same functionality can be replicated, but with less surface API. Plus, it makes it easier handling DPad buttons from the same place where you handle all your other buttons.

**Comment by AzyIsCool** [comment]
Ok, if someone really wanted it they would always make a extension. Got my approval.

File: documentation/proposals/Proposal - Input.md [hunk 3]
```diff
@@ -0,0 +1,390 @@
...
+- All modifications of states (i.e. functions) will go within the interface, whereas structs are used for read-only data.
+- **IMPLEMENTATION DETAIL/REQUIREMENT:** In implementations, once an IJoystick, IGamepad, or other IInputDevice object has been created, **it must not be destroyed until its IInputContext is disposed**. This is to preserve event bindings.
+- ButtonDown/Up and KeyDown/Up events (and similar) all fire once, when the key is first pushed down or up. I intend to champion a few extension packages to support more use cases, but I've kept this API pretty barebones.
+
+# Proposed API
+## Interfaces
+
+### IJoystick
+```cs
+public interface IJoystick : IInputDevice
```
Comment (devvoid) [inline documentation/proposals/Proposal - Input.md]: Should there maybe be a deadzone on joysticks too?

Comment (Perksey) [reply documentation/proposals/Proposal - Input.md]: Yes, good call!

**Pull request merged**